src/Compilers/Core/Portable/Serialization/ObjectReader.cs

```diff
@@ -138,8 +138,15 @@ public object ReadValue()
                     _cancellationToken,
                     TaskCreationOptions.LongRunning,
                     TaskScheduler.Default);
-                task.Wait(_cancellationToken);
-                value = task.Result;
+
+                // We must not proceed until the additional task completes. After returning from a read, the underlying
+                // stream providing access to raw memory will be closed; if this occurs before the separate thread
+                // completes its read then an access violation can occur attempting to read from unmapped memory.
+                //
+                // CANCELLATION: If cancellation is required, DO NOT attempt to cancel the operation by cancelling this
+                // wait. Cancellation must only be implemented by modifying 'task' to cancel itself in a timely manner
+                // so the wait can complete.
+                value = task.GetAwaiter().GetResult();
```

Comment:
Thanks stephen.  Sam's code  looks to be exaclty what we want then (i.e. throw exception directly, not an aggregate exception).  Preemptively cancel, but only have it respected on the next potential thread launch.  Thanks!